### PR TITLE
Add Ray implementation based on charm4py

### PR DESCRIPTION
**Pull request by adityapb**
Add ray implementation on charm4py

- Add an object store
- Add an argument `store` to `Future` which, if true, stores the value in a distributed object store instead of sending it to the creator PE
- Add a garbage collector to delete data from object store when no references to the future are left anywhere in the system
- Add a bool `is_ray` to the message header to differentiate between messages sent to ray actor methods/tasks from messages sent to charm4py method invocations
- On the receiving PE, if the message has `is_ray=True`, the message is buffered until the values of all futures passed as arguments are available locally

### CODE REVIEW

File: charm4py/charm.py [hunk 1]
```diff
@@ -153,6 +156,11 @@ def __init__(self):
         self.threadMgr = threads.EntryMethodThreadManager(self)
         self.createFuture = self.Future = self.threadMgr.createFuture
 
+        self.send_buffer = MessageBuffer()
```
Comment (adityapb) [inline charm4py/charm.py:159]: Add a comment that this isn't currently used

File: charm4py/charm.py [hunk 2]
```diff
@@ -166,6 +174,65 @@ def __init_profiling__(self):
...
+            local_f = LocalMultiFuture(num_returns - ready_count)
+            for f in not_local:
+                self.future_get_buffer[f.store_id] = (local_f, f)
+            result = local_f.get()
+            for f in not_local:
+                self.future_get_buffer.pop(f.store_id, None)
+            return ready_list + result
+        
+    def check_futures_buffer(self, obj_id):
+        if obj_id in self.future_get_buffer:
```
Comment (adityapb) [inline charm4py/charm.py:218]: Possible inefficiency

File: charm4py/charm.py [hunk 3]
```diff
@@ -166,6 +174,65 @@ def __init_profiling__(self):
...
+            local_f, fut = self.future_get_buffer.pop(obj_id)
+            local_f.send(fut)
+
+    def check_send_buffer(self, obj_id):
+        completed = self.send_buffer.check(obj_id)
+
+    def check_receive_buffer(self, obj_id):
```
Comment (adityapb) [inline charm4py/charm.py:225]: Avoid this call when ray.get on a future is called.
Replace future with object in the check function

File: charm4py/threads.py [hunk 4]
```diff
@@ -34,22 +37,35 @@ def __init__(self, fid, gr, src, num_vals):
         self.blocked = False  # flag to check if creator thread is blocked on the future
         self.gotvalues = False  # flag to check if expected number of values have been received
         self.error = None  # if the future receives an Exception, it is set here
+        if store:
+            self.store_id = (self.src << 32) + self.fid
```
Comment (adityapb) [inline charm4py/threads.py:41]: MAke number of bits for PE customizable

**Review by ritvikrao — APPROVED** [review]
Everything is set for the merge; remaining issues are minor and will be addressed later

**Pull request merged**